features/config/service/api/src/main/java/org/opennms/features/config/service/util/PropertiesCacheWithCm.java

```diff
@@ -0,0 +1,168 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2007-2021 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2021 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.features.config.service.util;
+
+import static org.opennms.features.config.service.util.PropertiesConversionUtil.mapToJsonString;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.opennms.core.sysprops.SystemProperties;
+import org.opennms.features.config.service.api.ConfigUpdateInfo;
+import org.opennms.features.config.service.api.ConfigurationManagerService;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+/**
+ * A direct replacement for org.opennms.core.utils.PropertiesCache.
+ * It operates on cm instead of a file.
+ */
+public class PropertiesCacheWithCm {
+
+  public static final String CHECK_LAST_MODIFY_STRING = "org.opennms.utils.propertiesCache.enableCheckFileModified";
+  public static final String CACHE_TIMEOUT = "org.opennms.utils.propertiesCache.cacheTimeout";
+  public static final int DEFAULT_CACHE_TIMEOUT = 3600;
+
+  protected static class PropertiesHolder {
```

Comment:
It is better to give a simple description about relations between PropertiesCacheWithCm & PropertiesHolder. Also is that means the get & setProperty in PropertiesHolder is not supposed to access directly?

Reply:
Yeah this is a good point. Its a pretty complicated class, I don't this it needs to be that way.
I originally copied it from PropertiesCache cache.
But we can simplify more.
I'll try

Reply:
simplified.